[Foundation] Coalesce duplicate String keys in bridged NSDictionary and NSSet

lorentey opened the pull request:
NSString has a stricter concept of equality than Swift’s String, so it is possible to construct NSDictionary/NSSet instances that contain duplicate keys according to Swift’s definition of string equality.

Currently, such NSDictionary/NSSet instances get bridged over as broken Swift collections, leading to various crashes and weird behavior. This change fixes set/dictionary bridging to coalesce such duplicate keys, by keeping a single value and discarding the rest. (This was the behavior of a previous bridging implementation in stdlib.)

We expect such cases are rare and mostly arise from bugs in processing Unicode data. Silently discarding data is not ideal; we should log a runtime warning when we come across duplicate keys. However, this is still a better option than keeping things as is, or forcing a trap.

rdar://problem/35995647

lorentey committed:
[Foundation] Coalesce duplicate keys in String-keyed NSDictionary and NSSet instances

NSString has a stricter concept of equality than Swift’s String, so it is possible to construct NSDictionary/NSSet instances that contain duplicate keys according to Swift’s definition of string equality. This change improves how we handle these cases by coalescing the duplicate keys (keeping a single value).

rdar://problem/35995647

lorentey commented:
@swift-ci please test

phausler commented on stdlib/public/SDK/Foundation/NSDictionary.swift:
```diff
@@ -80,6 +80,23 @@ extension Dictionary : _ObjectiveCBridgeable {
+      // string-keyed NSDictionaries may generate key collisions when bridged
+      // over to Swift. See rdar://problem/35995647
+      var dict = Dictionary(minimumCapacity: d.count)
+      d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
```
string keys are the most common form of NSDictionaries on iOS and macOS. This makes all bridges of NSDictionary (e.g. any API that vends them) O(N) recursively

phausler replied:
can we perhaps restrict the scope of the impact here to only strings that bridge across with varying composed/decomposed states?

lorentey replied:
Hm, I don't expect a major performance difference between this and the original `_DictionaryBuilder` case below; both are eagerly bridging dictionaries in O(n) steps. (Although there is certainly room for microoptimizations; e.g., `updateValue` could be replaced with a subscript set as long as we don't check its return value.)

Should we revisit the decision about eager/lazy bridging? My impression was that eager bridging is a win here overall; but I wasn't involved in the original experiments. Cc @airspeedswift

lorentey replied:
The mismatching hashes and String equality makes for quite a large gap between Swift and Foundation in this case. I believe we could implement a combination of lazy/deferred bridging where we pass lookups of single-line ASCII keys to `NSDictionary`, but we convert the entire dictionary on the first use of a nontrivial String key, or any time an indexing API is used. Deferred conversion requires mutating a potentially shared object, complicating implementation. I'm not entirely sure if the result would be an overall improvement for typical dictionary sizes and usage patterns, but it may be worth a try.

lorentey replied:
However, I think this bug should be fixed sooner than that. This PR does not change the performance of dictionary bridging, but it does work around a serious (if rare) unrecoverable issue. (String-keyed `NSDictionary` instances get bridged over to `[String: Foo]` using an O(n) copy both before and after this change.)

phausler reviewed (commented):
we really need to consider the perf impact here

phausler commented on validation-test/stdlib/Dictionary.swift:
```diff
@@ -3123,6 +3123,26 @@ DictionaryTestSuite.test("BridgedFromObjC.Nonverbatim.ArrayOfDictionaries") {
+  nsd.setObject(42, forKey: cafe1)
+  nsd.setObject(23, forKey: cafe2)
+  expectEqual(2, nsd.count)
+  expectTrue((42 as NSNumber).isEqual(to: nsd.object(forKey: cafe1)))
```
dont use the AppleScripting bridge version of `isEqual(to:)` use `isEqual()` instead :(

sadly we cannot change this one and it is really awful...

lorentey replied:
D'oh; fixed!

lorentey committed:
[test] Don’t use isEqual(to:)

lorentey committed:
Use Dictionary’s subscript setter rather than updateValue

It doesn’t matter which value we keep, and since we don’t currently check for duplicate values, there is no reason to retrieve the old value.

lorentey commented:
@swift-ci please benchmark

phausler commented:
We probably should have some additional perf tests for bridging but the results here quelch my concerns

lorentey commented:
OK, let's merge this so that we resolve the original issue.
I'll experiment with deferred bridging for Strings while I work on updating Dictionary internals.
